### PR TITLE
MapReduce implementation for OrderBy plugin

### PR DESCRIPTION
**Pull request by ananya-bhattacharya**

### CODE REVIEW

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ * Sort data based on composite keys specified by the user.
+ */
+public class CompositeKey implements Writable, WritableComparable<CompositeKey> {
+
+  private final Gson gson = new GsonBuilder().create();
+  private String structureRecordJSON;
+  private String sortFieldsJSON;
+  private String schemaJSON;
+  private Type mapType = new TypeToken<LinkedHashMap>() {
+  }.getType();
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: can it be on previous line?

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  private Type schemaType = new TypeToken<Schema>() {
+  }.getType();
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: same here

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  public CompositeKey() {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: You can remove this constructor if not doing anything specific

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    this.schemaJSON = dataInput.readUTF();
+  }
+
+  /**
+   * This comparator controls the sort order of the keys.
+   *
+   * @param other
+   * @return
+   */
+  public int compareTo(CompositeKey other) {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: rename other to compositeKeyToCompare or some relevant name.

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    Schema schema = gson.fromJson(schemaJSON, schemaType);
+    try {
+      StructuredRecord structuredRecord1 = StructuredRecordStringConverter.fromJsonString(this.structureRecordJSON, schema);
+      StructuredRecord structuredRecord2 = StructuredRecordStringConverter.fromJsonString(other.structureRecordJSON, schema);
+      LinkedHashMap<String, String> sortFields = gson.fromJson(sortFieldsJSON, mapType);
+
+      for (Map.Entry<String, String> entry : sortFields.entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+        if (key.contains("->")) {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: Put "->" in constant as using multiple time in file

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                                          structuredRecord2);
+      } else {
+        throw new IllegalArgumentException("Nested structured defined by -> should be of type RECORD");
+      }
+    } else {
+      return getSortOrder(key, value, structuredRecord1, structuredRecord2);
+    }
+  }
+
+  private int getSortOrder(String key, String value, StructuredRecord structuredRecord1, StructuredRecord structuredRecord2) {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: rename this method as look similar to getSortingOrder, can rename like comareRecords or any other relevant name

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+/**
+ * Sort data based on composite keys specified by the user.
+ */
+public class CompositeKey implements Writable, WritableComparable<CompositeKey> {
+
+  private final Gson gson = new GsonBuilder().create();
+  private String structureRecordJSON;
+  private String sortFieldsJSON;
+  private String schemaJSON;
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: you can have these fields final

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                                                  (Byte) structuredRecord2.get(key));
+    } else {
+      throw new IllegalArgumentException("Sorting can be performed only on the following CDAP data types:  " +
+                                           "String, Int, Long, Float, Double, Byte");
+    }
+    return res;
+  }
+
+  public String getStructureRecordJSON() {
+    return structureRecordJSON;
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: where are we using this getter?

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  }
+
+  public String getSortFieldsJSON() {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: same here

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    return sortFieldsJSON;
+  }
+
+  public String getSchemaJSON() {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: same here

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+ * the License.
+ */
+
+package co.cask.hydrator.plugin;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.Job;
+
+public class CompositeKeyDriver {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyDriver.java]: put some description for class. Check same for other classes

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      for (int i = 0; i < sortList.size() - 1; i++) {
+        String key = sortList.get(i);
+        String value = sortFields.get(key);
+        if (key.contains("->")) {
+          key = key.split("->")[0].trim();
+        }
+
+        Schema.Field field = structuredRecord1.getSchema().getField(key);
+        Schema.Type type = field.getSchema().getType();
+        if (type.equals(Schema.Type.STRING)) {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java]: nit:You can put if / else if block to check type, get sort order and return comparision integer in common class.

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+          throw new IllegalArgumentException("Sorting can be performed only on the following CDAP data types:  " +
+                                               "String, Int, Long, Float, Double, Byte");
+        }
+
+        if (res == 0) {
+          return res;
+        }
+      }
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java]: User logger to print stack trace.

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    return res;
+  }
+
+  private int getSortingOrder(String order) {
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyGroupComparator.java]: nit: can be put in common class

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyReducer.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+public class CompositeKeyReducer extends Reducer<CompositeKey, Text, CompositeKey, Text> {
+  @Override
+  protected void reduce(CompositeKey key, Iterable<Text> values, Context context) throws IOException,
+    InterruptedException {
+    for (Text value : values) {
+      context.write(key, value);
+    }
+  }
+}
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKeyReducer.java]: put new line at the end

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java (hunk 3 above)
Comment (ananya-bhattacharya) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: The default constructor is required by the comparator.
If not provided it gives initialization error

File: hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java (hunk 4 above)
Comment (ananya-bhattacharya) [inline hydrator-plugins/core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/CompositeKey.java]: These are overridden methods from WritableComparable .
Added @Override annotation to methods

File: hydrator-plugins/core-plugins/src/test/java/co/cask/hydrator/plugin/batch/aggregator/OrderByTest.java [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  }
+
+  @Test
+  public void test() throws IOException {
+    Gson gson = new GsonBuilder().create();
+    StructuredRecord structuredRecord1 = StructuredRecord.builder(INPUT)
+      .set("productName", "Angels & Demons Hardback")
+      .set("quantity", 1)
+      .set("price", 20.5)
+      .set("shippingAddress", StructuredRecord.builder(ADDRESS)
```
Comment (bhushankawadkar) [inline hydrator-plugins/core-plugins/src/test/java/co/cask/hydrator/plugin/batch/aggregator/OrderByTest.java]: nit: you can have a method to return address structure record by passing city, state and zip as parameters. You can use same method for shipping as well as billing address.

**Comment by bhushankawadkar** [comment]
LGTM

**Comment by ananya-bhattacharya** [comment]
https://github.com/ameyk/CASK-PSL/pull/6

**Comment by ananya-bhattacharya** [comment]
Closing PR due to reference issue on main repo.
Will create separate PR for this with the new review comments.

**Pull request closed without merging**